pubtools/pulplib/_impl/model/maintenance.py

```diff
@@ -0,0 +1,179 @@
+
+    last_updated_by = attr.ib(default=None, type=str)
+    """Person/party who updated the report last time"""
+
+    entries = attr.ib(default=attr.Factory(tuple), type=tuple)
+    """A tuple contains :class:`MaintenanceEntry` objects, indicates
+    which repositories are in maintenance mode and details.
+    If empty, then it means no repositories is in maintenance mode.
+    """
+
+    @classmethod
+    def from_data(cls, data):
+        """Create a new report with raw data
+
+        Args:
+            data (dict):
+                A dict containing a raw representation of the maintenance status.
+        Returns:
+            a new instance of ``cls``
+
+        Raises:
+            InvalidDataException
+                If the provided ``data`` fails validation against an expected schema.
+        """
+        try:
+            jsonschema.validate(instance=data, schema=cls._SCHEMA)
+        except jsonschema.exceptions.ValidationError as error:
+            LOG.exception("%s.from_data invoked with invalid data", cls.__name__)
+            raise InvalidDataException(str(error))
+
+        entries = []
+        for repo_id, details in data["repos"].items():
+            entries.append(MaintenanceEntry(id=repo_id, **details))
+
+        maintenance = cls(
+            last_updated=data["last_updated"],
+            last_updated_by=data["last_updated_by"],
+            entries=tuple(entries),
+        )
+
+        return maintenance
+
+    def _json(self):
+        """Output a json format report.
+
+        Returns:
+            A string contains the maintenance report in json format.
+        """
+        report = {
+            "last_updated": self.last_updated,
+            "last_updated_by": self.last_updated_by,
+            "repos": {},
+        }
+
+        for entry in self.entries:
+            report["repos"].update(
+                {
+                    entry.id: {
+                        "message": entry.message,
+                        "owner": entry.owner,
+                        "started": entry.started,
+                    }
+                }
+            )
+
+        return json.dumps(report, indent=4, sort_keys=True)
+
+    def add(self, repo_ids, **kwargs):
+        """Add entries to maintenance report and update the timestamp. Every
+        entry added to the report represents a repository in maintenace mode.
+
+        Args:
+            repo_ids (list[str]):
+                A list of repository ids. New entries with these repository ids will
+                be added to the maintenance report.
+
+            Optional keyword args:
+                message (str):
+                    Reason why put the repo to maintenance.
+                owner (str):
+                    Who set the maintenance mode.
+
+        """
+        message = kwargs.get("message") or "Maintenance mode is enabled"
+        owner = kwargs.get("owner") or OWNER
+
+        to_add = []
+        for repo in repo_ids:
```

Comment:
I think we need some kind of handling here in the case the caller asked to add a repo which is already in the report.  Right now, seems like that'd create duplicates (which can't actually be stored).  I guess it would make more sense to overwrite the earlier entry (e.g. updating the comment for a repo which is already in maintenance mode).

Slightly wacky implementation idea - what if you did this:

```
out = MaintenanceReport(last_updated_by=owner, entries=tuple(entries))
return MaintenanceReport._from_json(out._json())
```

i.e. you do a round-trip through the serialization format before you return the object:

- means you can't have bugs where you return a MaintenanceReport which can't actually be serialized or which "changes" during serialization (e.g. duplicate entries)
- based on implementation of `_json`, this will already de-duplicate entries without any extra code needed

Reply:
No, that's not the only possible way to get duplicates.  This class is public API and users of this library can create instances of it themselves with whatever entries they want.  In particular when working with the fake client, I think they would be expected to do that (i.e. the use-case is just set the maintenance mode on the fake Pulp server to some specific canned data before running a certain test).

So, I think the validator is still valuable to prevent anyone writing code which creates invalid objects.

Reply:
Resolving since we have validator now.